slsim/Halos/halos.py

```diff
@@ -0,0 +1,758 @@
+    :param cosmology: The cosmological model to apply, which defines the universe's geometry and expansion history, influencing the calculation of comoving volumes. This parameter is optional, with a default cosmology used if not specified.
+    :param sky_area: The area of the sky over which the volume calculations are to be applied, expressed in square degrees or as a solid angle, framing the scope of the astronomical observation or survey.
+    :type redshift_list: array_like
+    :type cosmology: astropy.Cosmology instance, optional
+    :type sky_area: `~astropy.units.Quantity`
+    :return: An array detailing the comoving volume corresponding to each redshift in the input list, providing a spatial context for the distribution of astronomical objects.
+    :rtype: array
+    """
+    dV_dz = (cosmology.differential_comoving_volume(redshift_list) * sky_area).to_value(
+        "Mpc3"
+    )
+    return dV_dz
+
+
+def halo_mass_at_z(
+    z,
+    m_min=None,
+    m_max=None,
+    resolution=None,
+    cosmology=None,
+    sigma8=0.81,
+    ns=0.96,
+    omega_m=None,
+):
+    """Sample halo masses at given redshift(s) using specified or default cosmological
+    parameters and mass range.
+
+    :param z: Redshift(s) at which to sample halo masses. Can be a single value or an
+        iterable of values.
+    :type z: float or iterable
+    :param m_min: Minimum halo mass in solar masses. Defaults to a predetermined value
+        if not specified.
+    :type m_min: float, optional
+    :param m_max: Maximum halo mass in solar masses. Defaults to a predetermined value
+        if not specified.
+    :type m_max: float, optional
+    :param resolution: Resolution of the computational grid. Defaults to a predetermined
+        value if not specified.
+    :type resolution: int, optional
+    :param cosmology: Cosmology instance to be used in calculations. Defaults to a
+        predetermined cosmology if not specified.
+    :type cosmology: astropy.cosmology instance, optional
+    :param sigma8: Sigma8 parameter for the power spectrum normalization. Defaults to
+        0.81 if not specified.
+    :type sigma8: float, optional
+    :param ns: Spectral index for the power spectrum. Defaults to 0.96 if not specified.
+    :type ns: float, optional
+    :param omega_m: Matter density parameter Omega_m. If not specified, the value from
+        the cosmology instance is used.
+    :type omega_m: float, optional
+    :return: List of sampled halo masses for each provided redshift. Each element
+        corresponds to the mass sampled at the respective redshift in the input.
+    :rtype: list
+    """
+
+    (
+        m_min,
+        m_max,
+        resolution,
+        cosmology,
+    ) = set_defaults(
+        m_min,
+        m_max,
+        resolution,
+        cosmology,
+    )
+    try:
+        iter(z)
+    except TypeError:
+        z = [z]
+
+    mass = []
+    if np.all(np.isnan(z)):
+        return [0] * len(z)
+    for z_val in z:
```

Comment:
I guess this operation is slow as it goes through a loop and draws one-by-one. Could there be an array-based method?